crates/ruff_python_formatter/src/comments/mod.rs

```diff
@@ -1,2 +1,274 @@
+
+//! Types for extracting and representing comments of a syntax tree.
+//!
+//! Most programming languages support comments allowing programmers to document their programs.
+//! Comments are different from other syntax because programming languages allow comments in almost any position,
+//! giving programmers great flexibility on where they can write comments:
+//!
+//! ```javascript
+//! /**
+//!  * Documentation comment
+//!  */
+//! async /* comment */ function Test () // line comment
+//! {/*inline*/}
+//! ```
+//!
+//! This flexibility makes formatting comments challenging because:
+//! * The formatter must consistently place comments so that re-formatting the output yields the same result,
+//!   and does not create invalid syntax (line comments).
+//! * It is essential that formatters place comments close to the syntax the programmer intended to document.
+//!   However, the lack of rules regarding where comments are allowed and what syntax they document requires
+//!   the use of heuristics to infer the documented syntax.
+//!
+//! This module tries to strike a balance between placing comments as closely as possible to their source location
+//! and reducing the complexity of formatting comments. It does so by associating comments per node rather than a token.
+//! This greatly reduces the combinations of possible comment positions, but turns out to be, in practice,
+//! sufficiently precise to keep comments close to their source location.
+//!
+//! Luckily, Python doesn't support inline comments, which simplifying the problem significantly.
+//!
+//! ## Node comments
+//!
+//! Comments are associated per node but get further distinguished on their location related to that node:
+//!
+//! ### Leading Comments
+//!
+//! A comment at the start of a node
+//!
+//! ```python
+//! # Leading comment of the statement
+//! print("test");
+//!
+//! [   # Leading comment of a
+//!     a
+//! ];
+//! ```
+//!
+//! ### Dangling Comments
+//!
+//! A comment that is neither at the start nor the end of a node.
+//!
+//! ```python
+//! [
+//!     # I'm between two brackets. There are no nodes
+//! ];
+//! ```
+//!
+//! ### Trailing Comments
+//!
+//! A comment at the end of a node.
+//!
+//! ```python
+//! [
+//!     a, # trailing comment of a
+//!     b, c
+//! ];
+//! ```
+//!
+//! ## Limitations
+//! Limiting the placement of comments to leading, dangling, or trailing node comments reduces complexity inside the formatter but means,
+//! that the formatter's possibility of where comments can be formatted depends on the AST structure.
+//!
+//! For example, *`RustPython`* doesn't create a node for the `/` operator separating positional only arguments from the other arguments.
+//!
+//! ```python
+//! def test(
+//!     a,
+//!     /, # The following arguments are positional or named arguments
+//!     b
+//! ):
+//!     pass
+//! ```
+//!
+//! Because *`RustPython`* doesn't create a Node for the `/` argument, it is impossible to associate any
```

Comment:
rustc wraps some of the operators with a span (and I'm considering this as well). But again, it's too late to add this.

https://doc.rust-lang.org/beta/nightly-rustc/rustc_ast/ast/type.BinOp.html

Reply:
That's interesting. That could be useful for my local python parser :)